sdk/scheduler/src/main/java/com/mesosphere/sdk/scheduler/TaskKiller.java

```diff
@@ -1,31 +1,128 @@
+    private static final Logger LOGGER = LoggerFactory.getLogger(TaskKiller.class);
+
+    private static final Duration killInterval = Duration.ofSeconds(5);
+    private static final Set<TaskID> tasksToKill = new HashSet<>();
+    private static final Object lock = new Object();
+
+    private static ScheduledExecutorService executor;
+    private static TaskKiller taskKiller = new TaskKiller();
+    private static SchedulerDriver driver;
+
+    private TaskKiller() {
+        startScheduling();
+    }
+
+    @VisibleForTesting
+    static void shutdownScheduling() throws InterruptedException {
+        executor.shutdownNow();
+        executor.awaitTermination(killInterval.toMillis(), TimeUnit.MILLISECONDS);
+    }
+
+    @VisibleForTesting
+    static void startScheduling() {
+        executor = Executors.newSingleThreadScheduledExecutor();
+        executor.scheduleAtFixedRate(
+                new Runnable() {
+                    @Override
+                    public void run() {
+                        killAllTasks();
+                    }
+                },
+                killInterval.toMillis(),
+                killInterval.toMillis(),
+                TimeUnit.MILLISECONDS);
+    }
 
-    public TaskKiller(SchedulerDriver driver) {
-        this.driver = driver;
+    public static void setDriver(SchedulerDriver schedulerDriver) {
+        driver = schedulerDriver;
     }
 
-    public void killTask(TaskID taskId) {
+    /**
+     * Calling this method will cause the referenced TaskID to be reliably killed.  An attempt to kill the task will be
+     * immediately made.  Continued periodic attempts will be made to kill the task until Mesos indicates that it has
+     * killed the Task or doesn't recongize the task.
+     *
+     * This is still not a guarantee that a task has been killed.  Mesos may not know about a particular TaskID at any
+     * given time.  See the {@link TaskCleaner} for the rest of the task killing system.
+     * @param taskId the TaskID of the Task to be killed.
+     */
+    public static void killTask(TaskID taskId) {
         // In order to update a podinstance its normal to kill all tasks in a pod.
         // Sometimes a task hasn't been launched ever but it has been recorded for
         // resource reservation footprint reasons, and therefore doesn't have a TaskID yet.
         if (taskId.getValue().isEmpty()) {
-            logger.warn("Attempted to kill empty TaskID.");
+            LOGGER.warn("Attempted to kill empty TaskID.");
             return;
         }
 
-        logger.info("Killing task: {}", taskId.getValue());
-        driver.killTask(taskId);
+        synchronized (lock) {
+            LOGGER.info("Enqueueing kill of task: {}", taskId.getValue());
+            tasksToKill.add(taskId);
+        }
+
+        killTaskInternal(taskId);
+    }
+
+    public static void update(Protos.TaskStatus taskStatus) {
+        // The race here is fine.  If we check for the set containing an element
```

Comment:
Instead of checking for presence and then removing, could just call remove() (within the lock) and then check the boolean return value to see if the removal was successful.

Reply:
I wanted to avoid acquiring the lock at all if I could.  It's probably not a big deal...

Reply:
I think you're right.  I'm going to just grab the lock.  It's too finicky.